From 0ab1d33dd8c2215a621324243975e81352cdb79e Mon Sep 17 00:00:00 2001
From: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Date: Tue, 26 Aug 2014 17:18:59 -0400
Subject: [input] alps: process_bitmap, fix counting of high point bits

Message-id: <1409073549-31548-9-git-send-email-benjamin.tissoires@redhat.com>
Patchwork-id: 89238
O-Subject: [PATCH RHEL7 08/18] Input: alps - process_bitmap: fix counting of high point bits
Bugzilla: 1107819
RH-Acked-by: Hans de Goede <hdegoede@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1107819
Upstream: since v3.17
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7882179
Tested by me on a Toshiba Portege R30 with a v7 Alps touchpad

commit 105affbfd588d5aec4171234051f7d589f7e62c1
Author: Hans de Goede <hdegoede@redhat.com>
Date:   Fri Jul 25 22:41:51 2014 -0700

    Input: alps - process_bitmap: fix counting of high point bits

    alps_process_bitmap was resetting the point bit-count as soon as it saw
    2 0 bits in a row. This means that unless the high point actually is at
    the end of the bitmap, it would always get its num_bits set to 0.

    Instead reset num_bits to 0 on a 0->1 transition, so that with > 2 fingers
    we only count the number of bits occupied by the highest finger.

    Signed-off-by: Hans de Goede <hdegoede@redhat.com>
    Signed-off-by: Dmitry Torokhov <dmitry.torokhov@gmail.com>

Signed-off-by: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/input/mouse/alps.c b/drivers/input/mouse/alps.c
index c6af590..5b35f4f 100644
--- a/drivers/input/mouse/alps.c
+++ b/drivers/input/mouse/alps.c
@@ -342,14 +342,13 @@ static void alps_get_bitmap_points(unsigned int map,
   if (bit) {
    if (!prev_bit) {
     point->start_bit = i;
+    point->num_bits = 0;
     (*fingers)++;
    }
    point->num_bits++;
   } else {
    if (prev_bit)
     point = high;
-   else
-    point->num_bits = 0;
   }
   prev_bit = bit;
  }
-- 
1.7.1